### PR TITLE
feat: JWT extractFrom headers and params

### LINKED ISSUE #2429: Add support JWT extraction from `headers` and `params`
*Description*:
Related to #2186 

Currently, I have a use case that demands multiple JWT extractions from the header (`Authorization Bearer`) or the cookies.

**The use case:**
My current org has invested time to move into Envoy Gateway as a replacement for Ingress, we have microservices coming from multiple entities in which they used different ways to extract the JWT, either from the header and/or the cookie. With the most recent version of Envoy Gateway, we can only satisfy JWT extraction from cookies but not from any other methods (e.g., headers or params).

[optional *Relevant Links*:]

### PR DESCRIPTION
**Pull request by ardikabs**
**What type of PR is this?**
Add other fields on `extractFrom` to support JWT extraction from custom headers or params.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Closes #2429

### CODE REVIEW

File: api/v1alpha1/jwt_types.go [hunk 1]
```diff
@@ -89,9 +89,35 @@ type ClaimToHeader struct {
...
 // JWTExtractor defines a custom JWT token extraction from HTTP request.
 type JWTExtractor struct {
+	// Headers represents a list of HTTP request headers to extract the JWT token from.
+	//
+	// +optional
+	Headers []JWTHeaderExtractor `json:"headers,omitempty"`
+
 	// Cookies represents a list of cookie names to extract the JWT token from.
 	// If specified, Envoy will extract the JWT token from the listed cookies and validate each of them.
 	// If any cookie is found to be an invalid JWT, a 401 error will be returned.
```
Comment (arkodg) [inline api/v1alpha1/jwt_types.go:99]: should we move this invalid JWT string to top level `JWTExtractor`, since its valid for all fields now ?

Comment (ardikabs) [reply api/v1alpha1/jwt_types.go:99]: that's right, will move it to the top

**Comment by ardikabs** [comment]
it seems the [job](https://github.com/envoyproxy/gateway/actions/runs/7499157956/job/20415690970?pr=2434#step:4:545) encountered an error that appears unrelated to the changes. Could you help here @arkodg?

```
--- FAIL: TestProvider (69.71s)
    --- FAIL: TestProvider/gatewayclass_with_parameters_ref (60.04s)
        kubernetes_test.go:199: 
            	Error Trace:	/home/runner/work/gateway/gateway/internal/provider/kubernetes/kubernetes_test.go:199
            	            				/home/runner/work/gateway/gateway/internal/provider/kubernetes/kubernetes_test.go:80
            	Error:      	Condition never satisfied
            	Test:       	TestProvider/gatewayclass_with_parameters_ref
```

**Comment by shawnh2** [comment]
> it seems the [job](https://github.com/envoyproxy/gateway/actions/runs/7499157956/job/20415690970?pr=2434#step:4:545) encountered an error that appears unrelated to the changes. Could you help here @arkodg?
> 
> ```
> --- FAIL: TestProvider (69.71s)
>     --- FAIL: TestProvider/gatewayclass_with_parameters_ref (60.04s)
>         kubernetes_test.go:199: 
>             	Error Trace:	/home/runner/work/gateway/gateway/internal/provider/kubernetes/kubernetes_test.go:199
>             	            				/home/runner/work/gateway/gateway/internal/provider/kubernetes/kubernetes_test.go:80
>             	Error:      	Condition never satisfied
>             	Test:       	TestProvider/gatewayclass_with_parameters_ref
> ```

seems like a flaky test

**Review by zhaohuabing — APPROVED** [review]
LGTM.

**Review by zirain — APPROVED** [review]

**Review by arkodg — APPROVED** [review]

**Pull request merged**